lang/src/org/partiql/lang/visitors/PartiQLVisitor.kt

```diff
@@ -1287,14 +1287,18 @@ internal class PartiQLVisitor(val ion: IonSystem, val customTypes: List<CustomTy
         scalarType("time_with_time_zone", listOfNotNull(precision))
     }
 
-    override fun visitTypeCustom(ctx: PartiQLParser.TypeCustomContext) = PartiqlAst.build {
+    override fun visitGenericType(ctx: PartiQLParser.GenericTypeContext) = PartiqlAst.build {
         val metas = ctx.symbolPrimitive().getSourceMetaContainer()
-        val customName: String = when (val name = ctx.symbolPrimitive().getString().toLowerCase()) {
-            in customKeywords -> name
-            in customTypeAliases.keys -> customTypeAliases.getOrDefault(name, name)
-            else -> throw ParserException("Invalid custom type name: $name", ErrorCode.PARSE_INVALID_QUERY)
+        when (val name = ctx.symbolPrimitive().getString().toLowerCase()) {
+            in customKeywords, in customTypeAliases.keys -> {
+                val customTypeAlias = customTypeAliases.getOrDefault(name, name)
+                customType_(SymbolPrimitive(customTypeAlias, metas), metas)
+            }
+            else -> {
+                val typeParameters = ctx.LITERAL_INTEGER().map { it.text.toLong() }
+                scalarType(name, typeParameters, metas)
+            }
         }
-        customType_(SymbolPrimitive(customName, metas), metas)
```

Comment:
Here we deal with custom types come first, since we still want to support the legacy `CustomType` API. The legacy `CustomType` API will be removed later as a separate goal of OTS, other than the pluggability goal.